### PR TITLE
Part of Issue1642: Enable use of Schnorr Signature type in Agora

### PR DESCRIPTION
**Pull request by hewison-chris**
This adds some missing changes required to use Signature in Agora.

### CODE REVIEW

File: source/agora/crypto/ECC.d [hunk 1]
```diff
@@ -168,7 +168,7 @@ public struct Scalar
 
     nothrow @nogc:
 
-    private this (typeof(this.data) data)
+    private this (typeof(this.data) data) pure inout
```
Comment (Geod24) [inline source/agora/crypto/ECC.d]: Tiny style nitpick, but type ctor go before attributes. `inout pure`

File: source/agora/crypto/Schnorr.d [hunk 2]
```diff
@@ -169,6 +163,7 @@ public Signature sign (T) (in Scalar privateKey, scope const auto ref T data)
 
 /// Sign with a given `r` (warning: `r` should never be reused with `x`)
 public Signature sign (T) (in Pair kp, in Pair r, scope const auto ref T data)
+    nothrow @nogc @safe
```
Comment (Geod24) [inline source/agora/crypto/Schnorr.d]: This looks unrelated (and unnecessary?)

File: source/agora/crypto/ECC.d (hunk 1 above)
Comment (linked0) [inline source/agora/crypto/ECC.d]: Can I get any information(like web pages) about `inout` used as a function attribute? It's always confusing to me. :)

File: source/agora/crypto/Schnorr.d [hunk 3]
```diff
@@ -128,6 +128,23 @@ public struct Signature
...
+    }
+
+    ///
+    @safe unittest
+    {
+        const signature = Signature(
+            Point("0x921405afbfa97813293770efd55865c01055f39ad2a70f2b7a04ac043766a693"),
+            Scalar("0x074360d5eab8e888df07d862c4fc845ebd10b6a6c530919d66221219bba50216"));
+        assert(signature.toString() == "0x921405afbfa97813293770efd55865c01055f39ad2a70f2b7a04ac043766a693074360d5eab8e888df07d862c4fc845ebd10b6a6c530919d66221219bba50216", signature.toString());
+    }
```
Comment (Geod24) [inline source/agora/crypto/Schnorr.d]: This allocates every time. We should use the `sink` overload instead.

File: source/agora/crypto/ECC.d (hunk 1 above)
Comment (hewison-chris) [reply source/agora/crypto/ECC.d]: This is quite a good description: https://forum.dlang.org/thread/d5ovc1$2rb$1@digitaldaemon.com

**Review by Geod24 — CHANGES_REQUESTED** [review]

File: source/agora/crypto/ECC.d (hunk 1 above)
Comment (Geod24) [reply source/agora/crypto/ECC.d]: Oh no, no, that's quite different :D

Comment (hewison-chris) [reply source/agora/crypto/ECC.d]: Also http://www.prowiki.org/wiki4d/wiki.cgi?FunctionParameterAttributes

Comment (hewison-chris) [reply source/agora/crypto/ECC.d]: As mentioned in Slack here is correct link:- https://dlang.org/spec/function.html#inout-functions

Comment (hewison-chris) [reply source/agora/crypto/ECC.d]: Updated

File: source/agora/crypto/Schnorr.d (hunk 2 above)
Comment (hewison-chris) [reply source/agora/crypto/Schnorr.d]: Reverted. I will add another commit to remove from the other overloaded functions which also have unnecessary attributes.

File: source/agora/crypto/Schnorr.d (hunk 3 above)
Comment (Geod24) [reply source/agora/crypto/Schnorr.d]: After https://github.com/bpfkorea/crypto/pull/25 you can just do:
```D
public void toString (scope void delegate(scope const(char)[]) @safe sink)
        const
{
    this.R.toString(sink);
    FormatSpec!char spec;
    spec.spec = 'x';
    this.s.toString(sink, spec);
}
```

Comment (hewison-chris) [reply source/agora/crypto/Schnorr.d]: Thanks, updated.

File: source/agora/crypto/ECC.d [hunk 4]
```diff
@@ -173,13 +173,29 @@ public struct Scalar
...
-        if (is(typeof(this.data = typeof(this.data)(param))))
     {
         this.data = typeof(this.data)(param);
     }
 
+    /// Construct from a dynamic array of the correct length
+    public this (ubyte[data.sizeof] param) inout pure
+    {
+        this.data = typeof(this.data)(cast(ubyte[data.sizeof])(param));
+    }
```
Comment (Geod24) [inline source/agora/crypto/ECC.d:186]: Why is this needed ? The previous template constraint should have covered this. And the `cast` should not be necessary, since `param` is already of type `ubyte[data.sizeof]`.

**Review by Geod24 — COMMENTED** [review]
AFAIK the last commit shouldn't be necessary. The others LGTM.

File: source/agora/crypto/ECC.d (hunk 4 above)
Comment (hewison-chris) [reply source/agora/crypto/ECC.d:186]: The unit test I added shows that the original constructor did not work for all cases. Also it was unclear how the constraint worked. If you prefer we don't include this last commit then can you show me how to get the unit test passing without it please.

Comment (Geod24) [reply source/agora/crypto/ECC.d:186]: The problem is:
```
        const from_ubytes = Scalar([57, 34, 14, 84, 18, 175, 101, 64, 121, 181, 212, 78, 23, 148, 180, 7, 9, 105, 237, 155, 78, 161, 191, 27, 97, 130, 209, 44, 202, 245, 208, 13]);
```

The compiler tells you what's wrong:
```
source/agora/crypto/ECC.d(35,35): Error: none of the overloads of this are callable using argument types (int[]), candidates are:
source/agora/crypto/ECC.d(206,13):        agora.crypto.ECC.Scalar.this(BitBlob!32LU data)
source/agora/crypto/ECC.d(219,12):        agora.crypto.ECC.Scalar.this(BitBlob!64LU param)
source/agora/crypto/ECC.d(212,12):        __ctor(T)(T param)
  with T = int[]
  must satisfy the following constraint:
       is(typeof(this.data = typeof(this.data)(param)))
```

That is, passing a literal such as `[1, 2, 3]` infers it as `int[]` not `ubyte[]`. However, with an explicit type, the compiler will try to infer a `ubyte[]`. Now while the explicit overload makes things better, I don't really see the use-case for literals except in unittests.

Comment (hewison-chris) [reply source/agora/crypto/ECC.d:186]: Well in the WellKnownKeys file we have thousands of use cases.

**Review by Geod24 — APPROVED** [review]

File: source/agora/crypto/ECC.d (hunk 4 above)
Comment (hewison-chris) [reply source/agora/crypto/ECC.d:186]: I pushed an update with the cast removed as it was not required. I would prefer to keep this extra constructor if you do not object as it will make it easier with the WellKnownKeys file generated from the vanity program.
